[INMA2472] Added whole summary for LINMA2472 + ressources (imgs) (#870)
Gp2mv3/Syntheses: src/q7/algodata-LINMA2472/summary/algodata-LINMA2472-summary.tex

\end{itemize}

We will now discover different models and explain the different properties in a little more detail as we go along.

\subsection{Models}

\subsubsection{Erdos-Renyi}

\begin{itemize}
    \item Fix number of nodes n
    \item Fix number of (undirected) edges m
    \item We have an uniform distribution on all graphs G(n,m)
\end{itemize}

We might be interested by a variant (\textbf{Gilbert}) where we fix p as the probability of being friend with someone. This is interesting because we can analyze the degree distribution :

\begin{quote}
Basically, if we have a probability of becoming friends with another, then we are in the case of a binomial distribution of mean $p(n-1) \approx pn $. So, a good approximation of that is a poisson where $\lambda = pn$ and $p \ll 1$. And by the CLT the poisson itself can be approximated by a gaussian if $pn \gg 1$.
\end{quote}

We discover some new properties: 
\begin{itemize}
    \item \textbf{Monotone increase property} : if you add edges, the properties are still valid.

>     \item \textbf{Every monotone property has a threshold function $p^*(n)$} : \\
>     If $\frac{p(n)}{p^*(n)} = 0$ then Prob(x) $\to 0$ and if conversely $p(n)$ is far above $p^*(n)$ we have the Prob(x) $\to 1$. \\
>     As an exemple : for ``at least one edge", we have the threshold function $p^*(n) = 1/n^2$. \\
>     $$\mathbb{E}(\sharp\{\text{edges}\}) = \Big( \begin{array}{c} n \\ 2 \end{array}\Big) p \sim n^2p \Rightarrow \begin{array}{c} 
>     p \ll p^* \quad \text{then} \quad \mathbb{E}(\sharp\{\text{edges}\}) \ll 1 \\
>     p \gg p^* \quad \text{then} \quad \mathbb{E}(\sharp\{\text{edges}\}) \gg 1
>     \end{array}
>     $$

Reviewer: ```suggestion
    \item \textbf{Every monotone property has a threshold function $p^*(n)$}:

    If $\frac{p(n)}{p^*(n)} = 0$ then Prob(x) $\to 0$ and if conversely $p(n)$ is far above $p^*(n)$ we have the Prob(x) $\to 1$.

    As an exemple : for ``at least one edge'', we have the threshold function $p^*(n) = 1/n^2$.

    \[ \mathbb{E}(\#\{\text{edges}\}) = \binom{n}{2} p \sim n^2p \Rightarrow \begin{cases}
    p \ll p^* & \text{ then } \mathbb{E}(\#\{\text{edges}\}) \ll 1 \\
    p \gg p^* & \text{ then } \mathbb{E}(\#\{\text{edges}\}) \gg 1
    \end{cases}
    \]
```

\end{itemize}

Note that we do not have heterogeneous degrees or density or many triangles : $\mathbb{E}(\sharp\{\bigtriangledown\}) = \Big( \begin{array}{c} n \\ 3 \end{array} \Big) p^3 = \Theta\big((pn)^3\big) = \Theta\big(\bar{d^3}\big)$

The last thing to remember about ER random graph is that the diameter and average distance are close to $ln(n)$ and $ln(pn) \sim ln(d)$. Basically this act kind of an `epidemic' with $log_d n$ generations to hit the whole component.
\subsubsection{Configuration}

In order to improve the previous models and make them more realistic, several approaches are discussed. 

\begin{itemize}
    \item \textbf{Fixing degree of each node}: one consider that the degree distribution is very wrong and impose a degree. 
    \item \textbf{Stub model}: we randomly wire defined stubs.
    \item \textbf{Random swaps}: taking an existing network and randomly swap edges.
\end{itemize}

This configuration model is similar in many aspects to ER random graphs, for example it still have an homogeneous density and a low diameter log n / log D where $D = \frac{\sum d_i^2}{\sum d_i}$ (but not homogeneous degree anymore).

\subsubsection{Watts-Strogatz (small world)}

> The basic idea behind this model is that most friends live nearby but it happens that some live far-away. We start with a circle and then randomly rewire each link with a probability p. A variant is adding links instead of editing them. \\

Reviewer: ```suggestion
The basic idea behind this model is that most friends live nearby but it happens that some live far-away. We start with a circle and then randomly rewire each link with a probability p. A variant is adding links instead of editing them.
```

\begin{figure}[h]
    \centering
    \includegraphics[width=.4\linewidth]{res/GN-Model.jpg}
\end{figure}

\newpage
\subsubsection{Stochastic block}
Even more general and flexible than the GN model, here we decompose the parts of the network into blocks, and for each block we say that there is a certain density. 

\begin{figure}[h]
    \centering
    \includegraphics[width=.8\linewidth]{res/SBM-Model.jpg}
\end{figure}

\subsection{Indicators of the structure}

\subsubsection{Basic}

\begin{itemize}
    \item \textbf{Degree distribution}
    \item \textbf{Clustering coefficient (recall)} :
    \begin{equation*}
        C_u = \frac{2T(u)}{d_u (d_u - 1)} \quad \text{where} \quad T(u) = \sharp\{\text{nodes linked to u} \}
    \end{equation*}
    \item \textbf{Betweenness centrality} : find certain edges where each shortest path between two zones must pass through. $O(n^3)$

    \item \textbf{Closeness centrality}: find the nodes that are closest to all the others in the network. $O(n^3)$
\end{itemize}

\subsubsection{K-Core}

> \textbf{Definition}: the k-core of a graph G(V,E) is its maximal subgraph G’(V’,E’) such that all nodes of G’ have a degree $\geq$ k \\

Reviewer: ```suggestion
\textbf{Definition}: the $k$-core of a graph $G(V,E)$ is its maximal subgraph $G’(V’,E’)$ such that all nodes of $G’$ have a degree $\geq k$.
```

> Scales linearly with the number of links in the network. Only applicable on undirected networks, and works best on unweighted networks. There exists a version for weighted networks, but it requires arbitrary thresholds. All nodes in the k+1-core are contained in the k-core. The largest possible core is therefore given by the maximal degree in the graph. The difference between the K+1-core and the K-core is called the K+1-shell. \\

Reviewer: ```suggestion
Scales linearly with the number of links in the network. Only applicable on undirected networks, and works best on unweighted networks. There exists a version for weighted networks, but it requires arbitrary thresholds. All nodes in the $k+1$-core are contained in the $k$-core. The largest possible core is therefore given by the maximal degree in the graph. The difference between the $K+1$-core and the $K$-core is called the $K+1$-shell.
```

The idea behind the algorithm is the following [Complexity O(m) and Space O(n+m)]: 

\begin{lstlisting}
while every node has not been pruned:
    to_prune = list(nodes of degree k)
    k_shell = empty_list()
    
    while to_prune is not empty:
        x = first elem of to_prune
        for each neighbor of x: 
            if it is not pruned:
                decrease by 1 its degree
                if its degree is now k: add to the list to_prune
        x is now pruned
        add x to k_shell
        increase total_pruned by 1
    
    store k_shell and do the same for k + 1
done
\end{lstlisting}

\subsubsection{S-Core}

S-core decomposition works exactly the same as k-core, except that we replace the degree of each node by its weighted degree s (the sum of the weights of its edges). If the link weights are diverse, the risk is, however, that we end up with each node in one core. A possible outcome for that is to create thresholds for defining cores, but unless there those can be linked to a concrete meaning, it’s difficult to have an objective way to separate those. Intervals are commonly used.

\subsection{Nodes with metadata}

\subsubsection{Assortativity}

> ``Birds of a feather flock together"\\

Reviewer: ```suggestion
``Birds of a feather flock together.''
```

\begin{itemize}

\item \textbf{Scalar node labels}

> Assume that we are in a simple graph and that every node has a scalar label (eg. age, degree, salary,..). We can pick a random edge and call the extremities X and Y : the assortativity coefficient of the label is Corr$(X,Y)$. (Newman)

Reviewer: ```suggestion
Assume that we are in a simple graph and that every node has a scalar label (eg. age, degree, salary\dots). We can pick a random edge and call the extremities X and Y: the assortativity coefficient of the label is Corr$(X,Y)$. (Newman)
```

\item \textbf{Binary labels}

If instead of labels we have binary categories (gender, smoker, ..) and want to know how many of our friend belongs to a certain one, we replace the label by -1 and 1 and compute, with m the nb of edges and $\mathcal{L}$ the label:
$$
BA_\mathcal{L} = 
\frac{
    \frac{
        |\text{Edges between same label}|
        }
        {m}
        - \sum_{\mathcal{L}} p^2_\mathcal{L}
    }
    {
    1-\sum_{\mathcal{L}} p^2_\mathcal{L}
    }
$$

\item \textbf{Categorical labels}: same as BA formula. 

\item \textbf{Non-simple graphs}: we consider weight as multiplicity of edge, and the probability of edge being chose is proprotional to weight. For directed graph, we must dinstinguish origin (X) and target (Y). (replacing $p^2_\mathcal{L}$ by $p_\mathcal{L}q_\mathcal{L}$)

\end{itemize}

> Be careful that the weighting is done on the number of degrees. Therefore, in the calculation the nodes of high degree will count more. This is not really a 'democratic' measure. Moreover, an assortativity of +1 is only possible for disconnected graphs and -1 for bipartite one. Plus, assortativity behaviour can be non uniform across large network.

Reviewer: ```suggestion
Be careful that the weighting is done on the number of degrees. Therefore, in the calculation the nodes of high degree will count more. This is not really a `democratic' measure. Moreover, an assortativity of +1 is only possible for disconnected graphs and -1 for bipartite one. Plus, assortativity behaviour can be non uniform across large network.
```

\subsection{Community detection}

> Community detection methods group nodes in highly-connected clusters, based on some objective functions. This allows to unfold the global organisation of nodes in clusters and make higher-order and finer analyses. However, it is not always possible or necessary to find a community.\\

Reviewer: ```suggestion
Community detection methods group nodes in highly-connected clusters, based on some objective functions. This allows to unfold the global organisation of nodes in clusters and make higher-order and finer analyses. However, it is not always possible or necessary to find a community.
```

    \begin{figure}[h]
        \centering
        \includegraphics[width=.7\linewidth, height=6cm]{res/Laplacian.jpg}
    \end{figure}
     
     The Normalised Laplacian is $NL = D^{-1/2} L D^{-1/2}$

    \item \textbf{Combinatorial methods}
    
    \item \textbf{Louvain method}: (Blondel, Guillaume, Lambiotte, Lefèvre, 2008)

    The inspiration for this method of community detection is the optimization of modularity as the algorithm progresses. One start from individual nodes as communities and for each node in turn, try merging it with other communities if it improves the modularity. Then each small community is grouped into one node and the first step is repeated. This method is \textbf{fast and random} (order in which nodes are considered changes) so we can test the robustness of the partition found by re-running and look distance/dissimilarity. For instance one use \textbf{Jaccard index}.
    
    \begin{figure}[h]
        \centering
        \includegraphics[width=.6\linewidth, height=5.5cm]{res/Louvain-algo.jpg}
    \end{figure}
\end{itemize}

\subsection{Diffusion processes}

\subsubsection{Mathematical epidemiology}

> Compartmental models: population is divided in compartments based on their health status. (Bernoulli in 1760 then $20^{th}$)\\

Reviewer: ```suggestion
Compartmental models: population is divided in compartments based on their health status. (Bernoulli in 1760 then $20^{th}$)
```
What did you mean by `$20^{th}$`?

They describe a variety of diseases, assume populations are homogeneously mixed and rely on parameters :
\begin{itemize}
    \itemsep0em
    \item $\beta$: average number of adequate contacts (i.e. contacts that make a transmission) for 1 individual for 1 time-step
    \item $1/\gamma$: average duration of the infectious period
    \item $1/\epsilon$: average duration of the latent period
    \item $1/\delta$: average duration of the passive immunity
\end{itemize}

\subsubsection{Linear threshold model (LTM)}

> Used in modelling of diffusion of innovation/ideas and `the spread of influence'. Type SI (stay infected). \\

Reviewer: ```suggestion
Used in modelling of diffusion of innovation/ideas and `the spread of influence'. Type SI (stay infected).
```

\subsubsection{Independent cascade model (ICM)}

\href{https://github.com/RomainGrx/LINMA2472-Homeworks/blob/main/miscellaneous/animations/cascade.gif}{Animation here.} With a set A0 of activated nodes at the start, at every step if a node u becomes activated it makes a single attempt to try to infect its neighbours v which succeeds with probability p. (independent of past trials). The process stops when no more nodes can be infected. We can perform link percolation: random removal of links in a graph, and study the size of the connected
components that are left.

\subsubsection{On transport networks}
\subsubsection{Influence of structure on diffusion of ideas}

The degree heterogeneity influences the spread of information on networks. The number of expected activated nodes at the first step is the same for both graphs: it is the average degree times the conversion probability. However, the degree of the nodes that are activated at the next steps have a higher expected degree than the average degree of the graph, particularly in heterogeneous networks.

\subsubsection{Limiting the spread of an epidemic}

> Limiting the spread of an epidemic can be worked on two levels:\\

Reviewer: ```suggestion
Limiting the spread of an epidemic can be worked on two levels:
```

> 
> \textbf{1}. Reducing the risk of contagion: adoption of guidelines\\
> \textbf{2}. Limiting the possible transmission paths by vaccination or limitation of social contacts. \\
> This is impacting the structure of the social network, and is related to percolation theory\\

Reviewer: ```suggestion
\begin{enumerate}
\item Reducing the risk of contagion: adoption of guidelines
\item Limiting the possible transmission paths by vaccination or limitation of social contacts.
\end{enumerate}
This is impacting the structure of the social network, and is related to percolation theory.
```

So the PCA is just a particular case of an auto-encoder with $\phi = \mathcal{V}_h$ and $\psi = \mathcal{V}_h^T$.

\vspace{1em}

So, the difference is that PCA computes linear projection on the space that maximises the variance of the data on each new axis when an auto-encoder computes non-linear projection (thanks to the activation functions) and tries to be able to reconstruct the input from the \textit{Feature Space}.

\subsection{Word embedding}

Word embedding is the process of mapping each word in a $\mathbb{R}^n$ space (embedding/feature space) such that similar words are close in the \textit{Feature Space} and opposite words will be far away.

\subsection{Word2Vec}

Given a text corpus the model trains a neural network to produce the word embedding based on its context in the document.

\begin{figure}[H]
    \centering
    \includegraphics[width=.5\textwidth]{res/context_corpus.png}
    \label{fig:context_corpus}
\end{figure}

\subsubsection{Continuous Bag of Words (CBOW)}

\begin{minipage}{.6\textwidth}
  
\textbf{Input} : context of around the word

\textbf{Output} : predict the target word

The purpose is to predict the target word from its context.

\end{minipage}

> \begin{minipage}{.4\textwidth}

Reviewer: ```suggestion
\begin{minipage}{.3\textwidth}
```

\begin{figure}[H]
    \centering

>     \includegraphics[width=.6\textwidth]{res/cbow.png}

Reviewer: ```suggestion
    \includegraphics[width=\textwidth]{res/cbow.png}
```

\end{figure}

\end{minipage}

\subsubsection{Skip-gram}

\begin{minipage}{.6\textwidth}
  
\textbf{Input} : target word

\textbf{Output} : predict the context of the target word

The purpose is to predict the context receiving the target word as input.

\end{minipage}

> \begin{minipage}{.4\textwidth}

Reviewer: ```suggestion
\begin{minipage}{.3\textwidth}
```

\begin{figure}[H]
    \centering

>     \includegraphics[width=.6\textwidth]{res/skip.jpeg}

Reviewer: ```suggestion
    \includegraphics[width=\textwidth]{res/skip.jpeg}
```

    \item \textbf{Auxiliary information} = Information known to an attacker.
    \item \textbf{Uniqueness} = : fraction of the dataset that is uniquely identified by the set A of quasi-identifiers
\end{itemize}

\subsection{Data pseudonymization and anonymization}
\begin{itemize}
    \item \textbf{Anonymization} = disassociating an individual’s record from their identity in a particular dataset. 
\end{itemize}

In short: if the data is anonymous it is not personal data anymore $\Rightarrow$ no consent, no purpose.

\begin{itemize}
    \item \textbf{Pseudonymization} = removal of direct identifiers (name, phone number, address, social security number…) and just give each person a unique id that cannot be linked to them.
\end{itemize}

Pay attention that pseudonymized data is not necessarly anonymous!

\subsubsection{How to pseudonymize data}
\begin{itemize}
    \item We could build a correspondence table between direct identifiers and random generated strings but:
    \begin{itemize}
        \item It quickly becomes cumbersome to save it and keep it up-to-date as new data arrives
        \item We need to make sure the strings are unique (collisions)
    \end{itemize}

>     \item We could use a secret formula (e.g. for phone number: pseudonymized ID=phone number*17+12345) but:

Reviewer: ```suggestion
    \item We could use a secret formula (e.g. for phone number: pseudonymized ID = phone number * 17 + 12345) but:
```

    \begin{itemize}
        \item If somebody knows the real phone numbers of a couple of people in the dataset it's easy to find back the formula (even with complex functions)
        \item And \textbf{collisions} are also an issue
    \end{itemize}

>     \item A good idea is to use 'cryptographic hash functions' but :

Reviewer: ```suggestion
    \item A good idea is to use `cryptographic hash functions' but :
```

    \item increase the amount of noise we’re adding with each new queries
\end{itemize}

But how can we know when to stop answering or how much noise to add? For this, we need to quantify how much information we are releasing with every answer. This is why Differential Privacy has been developed for.

\subsubsection{Differential Privacy}

%\todo[inline, backgroundcolor=orange!50]{TODO : finish. In the meantime here are some plic ploc facts.}
%\vspace{1em}

Differential privacy guarantees that the result of a certain query on the dataset
must be essentially the \textbf{same irrespective of the presence of any single individual}
in the dataset. $P(\text{output = y | Bob} \in D] \approx P(\text{output = y | Bob} \notin D] $.

\begin{quote}
    We define a dataset as a collection of rows, where each row is the record of
a different user. We will use $D$ to refer to the collection of all datasets. We say that
two different datasets D1 and D2 are \textbf{neighboring if they differ by exactly one
row}, which means $D1 = D2 \cup {r}$ or $D2 = D1 \cup {r}$ (where r is some row).
A mechanism is $\epsilon$-Differentially private if it satisfies this definition.
\end{quote}

> This is how Differential Privacy defines privacy. It is an “arbitrary” choice but one that gives \textbf{strong meaningful guarantees} and works well mathematically. Note: like for QBS, Differential Privacy considers everything sensitive and makes no distinction between quasi-identifiers and sensitive information.\\

Reviewer: ```suggestion
This is how Differential Privacy defines privacy. It is an “arbitrary” choice but one that gives \textbf{strong meaningful guarantees} and works well mathematically. Note: like for QBS, Differential Privacy considers everything sensitive and makes no distinction between quasi-identifiers and sensitive information.
```

> For DP, we however generally use \textbf{Laplace noise} instead of the Gaussian noise we used before. \\

Reviewer: ```suggestion
For DP, we however generally use \textbf{Laplace noise} instead of the Gaussian noise we used before.
```

> This is where the nicest feature from DP comes in: \textbf{composability}. Releasing the output of any two queries on a dataset protected by a $\epsilon$-DP mechanism is equivalent to releasing one query protected by a 2$\epsilon$-DP mechanism. We can then decide on an $\epsilon$, which we call the \textbf{privacy budget}, which then defines the total number of queries (any of them!) anyone can run on the dataset.\\

Reviewer: ```suggestion
This is where the nicest feature from DP comes in: \textbf{composability}. Releasing the output of any two queries on a dataset protected by a $\epsilon$-DP mechanism is equivalent to releasing one query protected by a 2$\epsilon$-DP mechanism. We can then decide on an $\epsilon$, which we call the \textbf{privacy budget}, which then defines the total number of queries (any of them!) anyone can run on the dataset.
```

> The \textbf{global sensitivity of a function f} captures the magnitude by which a single individual’s data can change the function f in the worst case, and therefore, the uncertainty in the response that we must introduce in order to hide anyone’s participation. Adding noise according to Lap($\Delta f / \epsilon$) prevent averaging attack!\\ 

Reviewer: ```suggestion
The \textbf{global sensitivity of a function f} captures the magnitude by which a single individual’s data can change the function f in the worst case, and therefore, the uncertainty in the response that we must introduce in order to hide anyone’s participation. Adding noise according to Lap($\Delta f / \epsilon$) prevent averaging attack!
```

> DP guarantees can be \textbf{extended to groups} of size k but at the cost of “\textbf{multiplying the noise}” by a factor k. So to protect a group of 10 people we’d have to go from Lap(10) to Lap(100).\\

Reviewer: ```suggestion
DP guarantees can be \textbf{extended to groups} of size k but at the cost of ``\textbf{multiplying the noise}'' by a factor k. So to protect a group of 10 people we’d have to go from Lap(10) to Lap(100).
```

> \textbf{Some other ideas} :\\

Reviewer: ```suggestion
\textbf{Some other ideas}:
```
